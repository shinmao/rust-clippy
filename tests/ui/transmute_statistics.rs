#![allow(unused)]
#![warn(clippy::transmute_statistics)]

pub struct struct_a {
    a: i8,
    b: i32,
    c: i8,
}

pub struct struct_b {
    a: i8,
    b: i32,
    c: i8,
}

fn main() {
    let a = 0u32;
    let a = &a as *const u32;
    let _: &u32 = unsafe { std::mem::transmute(a) };
    let _: &u32 = unsafe { std::mem::transmute::<_, &u32>(a) };

    let b: u8 = 10;
    let _ = unsafe { std::mem::transmute::<_, &u32>(&b) };

    let c: u32 = 10;
    let _ = unsafe { std::mem::transmute::<_, &u8>(&c) };

    let sa = struct_a { a: 10, b: 11, c: 12 };
    let sb: &struct_b = unsafe { std::mem::transmute(&sa) };
}
