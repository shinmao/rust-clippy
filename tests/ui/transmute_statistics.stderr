error: type `struct_a` should have an upper camel case name
  --> $DIR/transmute_statistics.rs:4:12
   |
LL | pub struct struct_a {
   |            ^^^^^^^^ help: convert the identifier to upper camel case: `StructA`
   |
   = note: `-D non-camel-case-types` implied by `-D warnings`

error: type `struct_b` should have an upper camel case name
  --> $DIR/transmute_statistics.rs:10:12
   |
LL | pub struct struct_b {
   |            ^^^^^^^^ help: convert the identifier to upper camel case: `StructB`

error: Here is transmute(*const uint>&uint=>warn)!
  --> $DIR/transmute_statistics.rs:19:28
   |
LL |     let _: &u32 = unsafe { std::mem::transmute(a) };
   |                            ^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: `-D clippy::transmute-statistics` implied by `-D warnings`

error: Here is transmute(*const uint>&uint=>warn)!
  --> $DIR/transmute_statistics.rs:20:28
   |
LL |     let _: &u32 = unsafe { std::mem::transmute::<_, &u32>(a) };
   |                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: Here is transmute(&uint>&uint=>warn)!
  --> $DIR/transmute_statistics.rs:23:22
   |
LL |     let _ = unsafe { std::mem::transmute::<_, &u32>(&b) };
   |                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: Here is transmute(&uint>&uint=>warn)!
  --> $DIR/transmute_statistics.rs:26:22
   |
LL |     let _ = unsafe { std::mem::transmute::<_, &u8>(&c) };
   |                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: Here is transmute(&struct>&struct=>warn)!
  --> $DIR/transmute_statistics.rs:29:34
   |
LL |     let sb: &struct_b = unsafe { std::mem::transmute(&sa) };
   |                                  ^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to 7 previous errors

